File: LeetCode/Reverse Linked List II.cpp
#include <iostream>
using namespace std;
/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        int idx = 0;
        ListNode * root = new ListNode(-1);
        root->next = head;
        ListNode * start;
        ListNode * cur = head;
        ListNode * prev = root;
        while (cur != NULL){
            idx++;
            if (idx >= m && idx <= n && m != n){
                if (idx == m){
                    start = prev;
                }
                if (idx == n){
                    if (start->next != NULL)
                        start->next->next = cur->next;
                    start->next = cur;
                }
                ListNode * post = cur->next;
                cur->next = prev;
                prev = cur;
                cur = post;
            }else{
                prev = cur;
                cur = cur->next;
            }

        }
        return root->next;
    }
};
int main(int argc, char const *argv[])
{
    ListNode * a = new ListNode(1);
    ListNode * b = new ListNode(2);
    ListNode * c = new ListNode(3);
    ListNode * d = new ListNode(4);
    ListNode * e = new ListNode(5);
    a->next = b;
    // b->next = c;
    c->next = d;
    d->next = e;
    Solution sol;
    ListNode *head = sol.reverseBetween(a, 1, 2);
    while(head!=NULL)
    {
        cout << head->val << endl;
        head = head->next;
    }
    return 0;
}

File: LeetCode/Reverse Linked List II.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    def reverseBetween(self, head, m, n):
        if m == n:
            return head
        step = 0
        tnode = head
        prenode = ListNode(-1)
        prenode.next = head
        root = prenode
        start = root

        while tnode != None:
            step += 1
            if (step < m) or (step > n):
                if step == m - 1:
                    start = tnode
                prenode = tnode
                tnode = tnode.next
            else:
                postnode = tnode.next
                tnode.next = prenode
                if step == n and start != None:
                    start.next.next = postnode
                    start.next = tnode
                prenode = tnode
                tnode = postnode

        return root.next

if __name__ == '__main__':
    s = Solution()
    a = ListNode(1)
    b = ListNode(2)
    c = ListNode(3)
    d = ListNode(4)
    e = ListNode(5)
    a.next = b
    b.next = c
    c.next = d
    d.next = e
    head = s.reverseBetween(a, 1, 5)
    while head != None:
        print head.val, "->",
        head = head.next
    print None

